sources/web/datalab/auth.ts

```diff
@@ -128,6 +128,17 @@ export function isSignedIn(): boolean {
   return (process.env.DATALAB_ENV != 'local' || fs.existsSync(appCredFile));
 }
 
+function setOauth2Client(parsed_url: any): void {
+  if (!oauth2Client) {
+    var OAuth2:any = google.auth.OAuth2;
+    // TODO(ojarjur): Ideally, we would get the host and port from the parsed_url
+    // rather than hard-coding them. However, the client ID and secret we are using
+    // are limited to localhost only. We should consider making this configurable, or
+    // using the OAuth flow for non-web applications.
+    oauth2Client = new OAuth2(clientId, clientSecret, 'http://localhost:8081/oauthcallback');
```

Comment:
I don't think the "init-method" approach will work. At least, I tried it and hit a roadblock.

The issue is that the port number we need to use is *not* the port number of the running server, but rather the port number that Docker is listening to.

So, for instance, we currently run the server at port 8080 *inside* the Docker container, however, we have Docker map port 8081 of the host machine to port 8080 of the Docker container.

That means the port we need to include in the OAuth callback is the 8081, not 8080. That, in turn, means that we can't set up the OAuth client at server setup. Instead, we have to wait for an incoming request so that we can parse the port number from the referer string.